test/integration/isolated/ingress_test.go

```diff
@@ -0,0 +1,157 @@
+	"github.com/kong/kubernetes-testing-framework/pkg/utils/kubernetes/generators"
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	netv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/e2e-framework/pkg/envconf"
+	"sigs.k8s.io/e2e-framework/pkg/features"
+
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/annotations"
+	"github.com/kong/kubernetes-ingress-controller/v3/test"
+	"github.com/kong/kubernetes-ingress-controller/v3/test/helpers/certificate"
+	"github.com/kong/kubernetes-ingress-controller/v3/test/integration/consts"
+	"github.com/kong/kubernetes-ingress-controller/v3/test/internal/testlabels"
+)
+
+func TestIngressGRPC(t *testing.T) {
+	const testHostname = "grpcs-over-ingress.example"
+
+	f := features.
+		New("essentials").
+		WithLabel(testlabels.NetworkingFamily, testlabels.NetworkingFamilyIngress).
+		WithLabel(testlabels.Kind, testlabels.KindIngress).
+		WithSetup("deploy kong addon into cluster", featureSetup(
+			withKongProxyEnvVars(map[string]string{
+				"PROXY_LISTEN": `0.0.0.0:8000 http2\, 0.0.0.0:8443 http2 ssl`,
+			}),
+		)).
+		Assess("deploying gRPC service exposed via Ingress", func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
+			cleaner := GetFromCtxForT[*clusters.Cleaner](ctx, t)
+			cluster := GetClusterFromCtx(ctx)
+			namespace := GetNamespaceForT(ctx, t)
+
+			t.Log("configuring secret")
+			tlsRouteExampleTLSCert, tlsRouteExampleTLSKey := certificate.MustGenerateSelfSignedCertPEMFormat(certificate.WithCommonName(testHostname))
+			secret := &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "secret-test",
+				},
+				Data: map[string][]byte{
+					"tls.crt": tlsRouteExampleTLSCert,
+					"tls.key": tlsRouteExampleTLSKey,
+				},
+			}
+
+			t.Log("deploying secret")
+			secret, err := cluster.Client().CoreV1().Secrets(namespace).Create(ctx, secret, metav1.CreateOptions{})
+			assert.NoError(t, err)
+			cleaner.Add(secret)
+
+			t.Log("deploying a minimal GRPC service exposed with Ingress via HTTPS")
+			ctx = deployGRPCServiceWithIngress(ctx, t, true)
+
+			t.Log("deploying a minimal GRPC service exposed with Ingress via HTTP")
+			return deployGRPCServiceWithIngress(ctx, t, false) // To test GRPC over HTTP.
+		}).
+		Assess("checking whether Ingress status is updated and gRPC traffic over HTTPS is properly routed", func(ctx context.Context, t *testing.T, c *envconf.Config) context.Context {
+			t.Log("verifying service connectivity via HTTPS and Ingress status update")
+			ctx = verifyGRPCServiceAndIngress(ctx, t, testHostname)
+
+			t.Log("verifying service connectivity via HTTP and Ingress status update")
+			return verifyGRPCServiceAndIngress(ctx, t, "")
+		}).
+		Teardown(featureTeardown())
+
+	tenv.Test(t, f.Feature())
+}
+
+func deployGRPCServiceWithIngress(ctx context.Context, t *testing.T, gRPCS bool) context.Context {
```

Comment:
We should try to avoid creating new test resource creation utilities in this style, where we have functions whose arguments control the resource parameters. The similar HTTP Ingress creator has been a source of pain when we need to to add some new tunable, since the signature change will require updates throughout the test codebase.

I expect this one would be especially difficult in that respect, since it both creates and deploys the resource. Functions that simply generate a resource and leave it up to their caller to deploy it at least allow flexibility by modifying the generated resource after instead of adding a new parameter.

Would you be open to something more similar to how KTF addons work, where there's a factory type (`TestGRPCIngress` or whatever) that offers methods to tune settings (`WithPath()` to set a path other than `/`, `WithProtocol()` to set either `grpc` or `grpcs`, etc.) a `Build()` method that creates the resource, applying overrides if set and defaults if not, and a `Deploy()` method to create it?

Reply:
+1 For splitting the resource generating and deploy steps.

Reply:
I agree in general, but in this particular case, it seems like an overkill. I followed the YAGNI principle here. Those helpers are only for this test to keep it DRY (check it for HTTP and HTTPS) and I don't see more use cases for them. Supporting gRPC with Ingress is a Kong-specific feature that is pretty limited, the aforementioned path configuration is not possible for it AFAIK.

Reply:
The Kong-side `path` ends up being the GRPC service/method information. Dunno if Ingress can actually shove that in--GRPCRoute can, but it's designed to have actual fields for those.

If we think it's a one-off, how much do we really need the separate create+deploy for each protocol? Can we inline the resource create+deploy+verify in the test itself like before?

It seems like the main point of the change is that we're now testing against a Kong instance with a plaintext `http2` listen also. I think we could still create a single `protocol=grpc,grpcs` Ingress and just call `grpcEchoResponds()` for both protocols to validate the plaintext and encrypted functionality fine. It doesn't look like we need a separate Ingress/Service/Deployment.

Reply:
Indeed for deployment and ingress single instances are enough. But two separate services have to be deployed each with proper protocol annotation. I made changes according to your comments, PTAL